### PR TITLE
[ci] Manually merge code after checkout

### PR DESCRIPTION
**Pull request by driazati**
This is step 1 of 2 to fix the issue where Jenkins schedules a PR rebuild on non-code changes like editing the title. When the title is changed, GitHub sends an `pull_request` webhook to GitHub. Jenkins would ordinarily look at the webhook, query the PR and checks if the commit hash for the build has already been built. However, since we have it set in the Jenkins job to merge with main before this check occurs, the commit hash is different nearly every time. Disabling that setting fixes the issue, but we still need to merge the code with `main` to ensure that the build is valid if it completes successfully.

Also see the Jenkins configuration change here: https://github.com/tlc-pack/ci/pull/22

cc @areusch

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -100,6 +100,10 @@ def init_git() {
     script: './tests/scripts/task_show_node_info.sh',
     label: 'Show executor node info',
   )
+  sh (
```
Comment (areusch) [inline Jenkinsfile:103]: i think this might mean that different parts of the build use different merge commits, depending on when this runs, right?

Comment (driazati) [reply Jenkinsfile:117]: fixed in the update, it fetches origin once at the beginning and checks out that same commit in further stages

File: Jenkinsfile [hunk 2]
```diff
@@ -100,6 +101,23 @@ def init_git() {
     script: './tests/scripts/task_show_node_info.sh',
     label: 'Show executor node info',
   )
+
+  // Determine merge commit to use for all stages
+  sh(
+    script: 'git fetch origin main',
+    label: 'Fetch upstream',
+  )
+  if (upstream_revision == null) {
```
Comment (areusch) [inline Jenkinsfile:110]: q: what about race conditions with this approach?

Comment (driazati) [reply Jenkinsfile:110]: Right now it's ok since this should happen during the `Sanity Check` stage which gates parallel execution on a sequential job

**Review by areusch — APPROVED** [review]

**Pull request merged**